src/defrag.c

```diff
@@ -761,16 +756,20 @@ void defragScanCallback(void *privdata, const dictEntry *de) {
     server.stat_active_defrag_scanned++;
 }
 
-static void defragKvstoreDefragScanCallBack(dict **d) {
-    dict *newd;
-    /* handle the dict struct */
-    if ((newd = activeDefragAlloc(*d)))
-        *d = newd;
-    dictDefragTables(*d);
-}
+static void defragPubsubCallback(dict **d) {
+    /* defrag the dict tables */
+    dictDefragTables(d);
 
-void activeDefragKvstore(kvstore *kvs) {
-    kvstoreDictLUTDefrag(kvs, defragKvstoreDefragScanCallBack);
+    /* Defragment the pubsub clients dictionary which is the 'value' part of the dictionary.
+     * We don't defrag the keys because they are also referenced by c->pubsub(shard)_channels. */
```

Comment:
can't we still defrag these keys, and locate the client and update the reference?
i think we can and we should.

Reply:
It's not easy to backtrack which client a channel string object is referenced by, for example, if there are 100 clients subscribed to channel A, then A in the pubsub dict is only referenced by one of the 100 clients.
Maybe we can dereference it by allocating a channel name when adding to the pubsub dict.
Another option is to defrag the pubsub dict by traversing the clients, feel that it's not a good idea.

Reply:
Already used the new way in https://github.com/redis/redis/pull/13058#discussion_r1497667248